### ISSUE TITLE
BlockingIOError in the handler

### ISSUE THREAD

**Issue by cyberic99** [Delgan/loguru#167]
Hello

I have encountered this errors quite often, but I do not know yet how to reproduce it...

Here is the error I get:
```
--- Logging error in Loguru Handler #1 ---
--- Logging error in Loguru Handler #1 ---
--- Logging error in Loguru Handler #1 ---
--- Logging error in Loguru Handler #1 ---
--- Logging error in Loguru Handler #1 ---
--- Logging error in Loguru Handler #1 ---
--- Logging error in Loguru Handler #1 ---
Record was: {'elapsed': datetime.timedelta(seconds=2054, microseconds=690483), 'exception': None, 'extra': {'log_category': 'PROCESS '}, 'file': '<string>', 'function': 'log_process', 'level': 'INFO', 'line': 1, 'message': 'Start thread 140162092386048 (pid=21517)', 'module': '<string>', 'name': 'Log', 'process': '21517', 'thread': '140162092386048', 'time': datetime(2019, 11, 17, 0, 52, 22, 717842, tzinfo=datetime.timezone(datetime.timedelta(seconds=3600), 'CET'))}
0:34:14.690483 [21517] PL/main         PROCESS  Start thread 140162092386048 (pid=21517)
0:34:14.690841 [21517] PL/main                  Traceback (most recent call last):

--- Logging error in Loguru Handler #1 ---
0:34:14.691027 [21517] PL/main                    File "/home/eric/.local/lib/python3.7/site-packages/loguru/_handler.py", line 144, in emit
    self._writer(str_record)

0:34:14.691188 [21517] PL/main                    File "/home/eric/.local/lib/python3.7/site-packages/loguru/_logger.py", line 696, in writer
    write(m)

0:34:14.691365 [21517] PL/main                  BlockingIOError: [Errno 11] write could not complete without blocking

--- End of logging error ---
```

One of my handlers is sending packets through UDP (on the same host), but it could be blocking

Do you know how I could debug this issue if it happens again?

Moreover, I *DO* catch exceptions when sending, so I am not sure of the exact cause of this exception...

thx

Eric

**Comment by Delgan** [comment]
Hi @cyberic99.

Is is possible to see the code you're using to configure your handler please? What do your UDP sink and it's `.write()` method look like? What are the options used while calling `.add()`?

**Comment by cyberic99** [comment]
My handler is used to send my program logs to [Open Stage Control](https://openstagecontrol.ammd.net/), using the OSC protocol

Here is the handler, as is.

```
class OpenStageControlLogSender:
    def __init__(self, dmn):
        self.dmn = dmn
        self.do_exit = False
        svces = dmn.find_services_with_name("Open Stage Control", sa=self.dmn.services_available)
        clients = [get_osc_client(svce) for svce in svces]
        if clients is None or len(clients) == 0:
            return
        self.client = clients[0]
        self.lines = []
        self.t = Thread(target=self.send_logs)
        self.t.daemon = True
        self.t.start()

    def __del__(self):
        self.lines = []

    def send_logs(self):
        while True:
            time.sleep(0.5)

            # FIXME: nobody sets it to True for now
            if self.do_exit:
                break

            if len(self.lines) > 0:
                osc_msg = osc_message_builder.OscMessageBuilder(address="/LOG_LINE")
                osc_msg.add_arg("\n".join(self.lines[-4096:]))
                self.lines = []
                m = osc_msg.build()
                try:
                    self.client.send(m)
                except Exception:
                    # We sometimes get an OSError: [Errno 90] Message too long
                    # When trying to send too much over UDP
                    pass

    def write(self, buf):
        self.lines.extend(buf.splitlines())
```
I am calling add() like this:
```
logger.add(OpenStageControlLogSender(self), colorize=False, format=LOG_FMT, level="INFO")
```

thx

**Comment by Delgan** [comment]
Thanks, it helps to understand!

It's quite surprising that you are seeing an error, because Loguru only wraps the `write()` method, and yours does almost nothing!

I'm starting to think... Is that really a problem in your `OpenStageControlLogSender`? Can you confirm that the handler `#1` is the OSC's one? This can be checked by printing the number returned by `.add()`. Are you also re-adding an handler logging to `sys.stdout`?

Basically, this error seems to happen when there is too much data to be written, and some internal buffer can't handle it: [Logging chokes on BlockingIOError: write could not complete without blocking](https://stackoverflow.com/questions/54185874/logging-chokes-on-blockingioerror-write-could-not-complete-without-blocking)

**Comment by cyberic99** [comment]
In fact I have one or 2 additional handlers:

a console logger:
```
logger.add(sys.stdout, colorize=colorize, format=LOG_FMT, level="INFO")
```
and a file logger (``logto`` contains the file name)
```
logger.add(logto, mode="w", colorize=False, format=LOG_FMT, level="INFO")
```
Also, I do that:
```
def myprint_loguru(*objects, sep=None, end=None, file=None, flush=False):
    msg = ' '.join(map(str, objects))
    logger.info(msg)

builtins.print = myprint_loguru
```

**Comment by Delgan** [comment]
Do you know which of these handlers is assigned the `#1` identifier (the one seen in the error message)? This should be the first one you `.add()` right after calling `.remove()`. The handler identifier is returned by the `.add()` method, so you can also print it to be sure. I would like to know which of yours handlers is causing the error.

Re-assigning `builtins.print` is not a problem as far as I know.

**Comment by cyberic99** [comment]
The logger number one is the one used there:

logger.add(sys.stdout, colorize=colorize, format=LOG_FMT, level="INFO")

I forgot to say that the issue I described seem to happen after a ``subprocess.Popen()`` with ``tdout=subprocess.PIPE, stderr=subprocess.STDOUT`` and... there must be a typo because I also have ``bufsize=1``...
yes, 1, not -1...

Could it be the cause of the problem?

**Comment by Delgan** [comment]
So, this is not related to your UDP handler. This is a problem while writing to `stdout`. 

Starting a `Popen` process with `bufsize=1` means it's line buffered. I don't think it's causing trouble. However, maybe that the sub-process stdout pipe is somehow conflicting with your handler... I don't know.

I'm sorry I can't help you much. I never encountered such issue. This does not seem related to Loguru itself. It seems it happens if too much data is outputed... Maybe try to reproduce it without involving logging, just using `sys.stdout.write()`...

**Comment by Delgan** [comment]
I'm closing this issue because   don't think it's a problem caused by Loguru itself.

If you have more information or need more help, please ask me. :+1:
